### PR TITLE
Implement EMA of the model

### LINKED ISSUE #1004: Exponential Moving Average (EMA)
### Is your feature request related to a problem? Please describe.

Implement EMA of the model weights, this should improve performance and form a good basis for any contrastive approaches a la DINO in the future

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Description

Standard machine learning trick to improve results, common in Diffusion in particular!

## Issue Number

Closes #1004 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

I tested with pre-training on ERA5 and ERA5+NPPATMS+SYNOP on 2 GPUs

### CODE REVIEW

**Comment by sophie-xhonneux** [comment]
This is based on FSDP2, still very much in draft mode and only tested for 1 GPU, so the multi-GPU case is a ToDo, hence the draft mode.

File: config/default_config.yml [hunk 1]
```diff
@@ -129,7 +129,7 @@ weight_decay: 0.1
 norm_type: "LayerNorm"
 nn_module: "te"
 
-start_date: 197901010000
+start_date: 199001010000
```
Comment (clessig) [inline config/default_config.yml]: Can we revert this

File: src/weathergen/model/ema.py [hunk 2]
```diff
@@ -8,52 +8,51 @@
...
+        self.ema_net = empty_net
+        self.is_model_sharded = is_model_sharded
+
+        self.reset()
 
     @torch.no_grad()
     def reset(self):
-        for p_net, p_ema in zip(self.net.parameters(), self.ema.parameters(), strict=False):
-            p_ema.copy_(p_net)
+        self.ema_net.to_empty(device="cuda")
```
Comment (clessig) [inline src/weathergen/model/ema.py]: Could we add a description to the function of what it does and why the copy via state_dict is needed

File: src/weathergen/model/ema.py [hunk 3]
```diff
@@ -8,52 +8,51 @@
...
     Taken and modified from https://github.com/NVlabs/edm2/tree/main
     """
 
     @torch.no_grad()
-    def __init__(self, net, halflife_steps=float("inf"), rampup_ratio=0.09):
-        self.net = net
+    def __init__(
+        self, net, empty_net, halflife_steps=float("inf"), rampup_ratio=0.09, is_model_sharded=False
+    ):
+        self.og_net = net
```
Comment (clessig) [inline src/weathergen/model/ema.py]: I would prefer `origina_net` or `base_net` or something. `og` is not self-explanatory

File: src/weathergen/model/ema.py [hunk 4]
```diff
@@ -8,52 +8,51 @@
...
             p_ema.lerp_(p_net, 1 - beta)
 
     @torch.no_grad()
     def forward_eval(self, *args, **kwargs):
-        self.ema.eval()
-        out = self.ema(*args, **kwargs)
-        self.ema.train()
+        self.ema_net.eval()
+        out = self.ema_net(*args, **kwargs)
+        self.ema_net.train()
```
Comment (clessig) [inline src/weathergen/model/ema.py]: Why reset it to train? The EMA network will never be used in training mode?

File: src/weathergen/train/trainer.py [hunk 5]
```diff
@@ -295,6 +297,17 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
         if cf.compile_model:
             self.model = torch.compile(self.model, dynamic=True)
 
+        cf.ema = True
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: TODO: move to config

File: src/weathergen/train/trainer.py [hunk 6]
```diff
@@ -580,6 +593,12 @@ def train(self, epoch):
             # update learning rate
             self.lr_scheduler.step()
 
+            # EMA update
+            self.ema_net.update(
```
Comment (clessig) [inline src/weathergen/train/trainer.py]:
```
if self.cf.ema : 
  # update EMA net
```

File: src/weathergen/train/trainer.py [hunk 7]
```diff
@@ -596,7 +615,8 @@ def train(self, epoch):
             if bidx % self.train_log_freq.checkpoint == 0 and bidx > 0:
                 self.save_model(-1)
 
-            self.cf.istep += cf.batch_size_per_gpu
+            # Shouldn't this either be 1 or cf.batch_size_per_gpu * self.original
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Remove (istep counts the optimizer steps, not the samples processed (in which case we would have to have cf.batch_size_per_gpu ... ) or something else)

File: src/weathergen/train/trainer.py [hunk 8]
```diff
@@ -295,6 +297,17 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
         if cf.compile_model:
             self.model = torch.compile(self.model, dynamic=True)
 
+        cf.ema = True
+        if cf.ema:
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: This parameter should be called `validate_with_ema` or similar to make clear where the EMA is used.

File: src/weathergen/model/ema.py [hunk 9]
```diff
@@ -8,52 +8,51 @@
...
+        self.is_model_sharded = is_model_sharded
+
+        self.reset()
 
     @torch.no_grad()
     def reset(self):
-        for p_net, p_ema in zip(self.net.parameters(), self.ema.parameters(), strict=False):
-            p_ema.copy_(p_net)
+        self.ema_net.to_empty(device="cuda")
+        maybe_sharded_sd = self.og_net.state_dict()
```
Comment (clessig) [inline src/weathergen/model/ema.py]: What does `maybe` mean hear?

File: src/weathergen/model/ema.py [hunk 10]
```diff
@@ -8,52 +8,51 @@
...
+        # this copies correctly tested in pdb
+        mkeys, ukeys = self.ema_net.load_state_dict(maybe_sharded_sd, strict=False, assign=False)
```
Comment (clessig) [inline src/weathergen/model/ema.py]: Why strict=False

File: src/weathergen/model/ema.py [hunk 11]
```diff
@@ -8,52 +8,51 @@
...
 
     @torch.no_grad()
     def update(self, cur_steps, batch_size):
         halflife_steps = self.halflife_steps
         if self.rampup_ratio is not None:
             halflife_steps = min(halflife_steps, cur_steps / 1e3 * self.rampup_ratio)
         beta = 0.5 ** (batch_size / max(halflife_steps * 1e3, 1e-6))
-        for p_net, p_ema in zip(self.net.parameters(), self.ema.parameters(), strict=False):
+        for p_net, p_ema in zip(self.og_net.parameters(), self.ema_net.parameters(), strict=False):
```
Comment (clessig) [inline src/weathergen/model/ema.py]: Why strict=False

File: src/weathergen/train/trainer.py [hunk 12]
```diff
@@ -295,6 +297,17 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
...
+            meta_ema_net = self.init_model_and_shard(cf, devices)[0]
+            self.ema_net = EMANeuralNet(
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Can we name this `self.ema_model` for consisteny (or use `self.net` instead `self.model`)

File: src/weathergen/model/ema.py (hunk 4 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/ema.py]: it may be I guess, it is just good practice imho

File: src/weathergen/train/trainer.py (hunk 5 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: ah yes, forgot this

File: src/weathergen/model/ema.py [hunk 13]
```diff
@@ -8,52 +8,51 @@
...
     def reset(self):
-        for p_net, p_ema in zip(self.net.parameters(), self.ema.parameters(), strict=False):
-            p_ema.copy_(p_net)
+        self.ema_net.to_empty(device="cuda")
+        maybe_sharded_sd = self.og_net.state_dict()
+        # this copies correctly tested in pdb
+        mkeys, ukeys = self.ema_net.load_state_dict(maybe_sharded_sd, strict=False, assign=False)
 
     @torch.no_grad()
     def update(self, cur_steps, batch_size):
```
Comment (clessig) [inline src/weathergen/model/ema.py]: Why is `cur_steps` plural?

File: src/weathergen/model/ema.py (hunk 9 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/ema.py]: the state dict may or may not contain sharded tensors, trying to convey some kind of type information that python fundamentally lacks

File: src/weathergen/model/ema.py (hunk 10 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/ema.py]: because it is the same in load_model and there may be new modules I guess?

File: src/weathergen/model/ema.py (hunk 11 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/ema.py]: see above

File: src/weathergen/model/ema.py (hunk 13 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/ema.py]: sounded right to me, but can change

**Comment by sophie-xhonneux** [comment]
I am addressing your comments, can you already approve so I can merge after?

File: src/weathergen/model/ema.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/ema.py]: ah to me og is a common enough shortform of original

File: src/weathergen/model/ema.py (hunk 10 above)
Comment (clessig) [reply src/weathergen/model/ema.py]: In load_model it's False so that we can add encoder/decoders a posteriori. I don't think it should ever differ here, but that's why I think it should be true so that we get warned if something happens that we didn't think about.

File: src/weathergen/model/ema.py (hunk 4 above)
Comment (clessig) [reply src/weathergen/model/ema.py]: Ok

File: src/weathergen/train/trainer.py (hunk 7 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: adding this to the design doc on self.cf.istep
